### PR TITLE
Fixed foreign elements (svg, mathml)

### PR DESCRIPTION
**Pull request by jaytaph**
This PR is a merge between @CharlesChen0823 PR and my foreign elements PR. 

It was a bit tricky as a lot has been moved around, but all the tests we assumed should pass, indeed pass.

I've also added ALL test dat files to the test and see how well they worked. Turns out a lot of them passes without any issues. All tests currently commented in the code are the files where at least 1 test does not pass (i haven't tested them manually or added them to the disabled cases).

### CODE REVIEW

**Comment by jaytaph** [comment]
@CharlesChen0823 Could you check my PR as well?

File: tests/tree_construction.rs [hunk 1]
```diff
@@ -29,6 +39,18 @@ const DISABLED_CASES: &[&str] = &[
     "</body><frameset></frameset>",
     // tests7.dat
     "<body>X</body></body>",
+    // tests8.dat
```
Comment (CharlesChen0823) [inline tests/tree_construction.rs:42]: tests8.dat can remove, I think all had pass.

File: tests/tree_construction.rs [hunk 2]
```diff
@@ -29,6 +39,18 @@ const DISABLED_CASES: &[&str] = &[
...
+    "x<table><table>x",
+    // tests10.dat
+    "<!DOCTYPE html><body xlink:href=foo xml:lang=en><svg><g xml:lang=en xlink:href=foo />bar</svg>",
+    "<div><svg><path><foreignObject><math></div>a",
+    "<div><svg><path><foreignObject><p></div>a",
+    "<!DOCTYPE html><p><svg><desc><p>",
+    "<!DOCTYPE html><p><svg><title><p>",
+    "<svg><script></script><path>",
+    // tests24.dat
+    "<!DOCTYPE html>&ThickSpace;A",
```
Comment (CharlesChen0823) [inline tests/tree_construction.rs:53]: tests24.dat can also remove, I think all had pass.

File: tests/tree_construction.rs [hunk 3]
```diff
@@ -6,6 +6,16 @@ use test_case::test_case;
 const DISABLED_CASES: &[&str] = &[
     // tests2.dat
     "<!DOCTYPE html>X<p/x/y/z>",
```
Comment (CharlesChen0823) [inline tests/tree_construction.rs:8]: tests2.dat can also remove.

Comment (jaytaph) [reply tests/tree_construction.rs]: removed

File: tests/tree_construction.rs (hunk 2 above)
Comment (jaytaph) [reply tests/tree_construction.rs]: removed

File: tests/tree_construction.rs (hunk 1 above)
Comment (jaytaph) [reply tests/tree_construction.rs]: removed

File: tests/tree_construction.rs [hunk 4]
```diff
@@ -42,16 +56,60 @@ lazy_static! {
...
 #[test_case("tests2.dat")]
 #[test_case("tests3.dat")]
+#[test_case("tests4.dat")]
 #[test_case("tests5.dat")]
 #[test_case("tests6.dat")]
 #[test_case("tests7.dat")]
 #[test_case("tests8.dat")]
+// #[test_case("tests9.dat")]
+#[test_case("tests10.dat")]
+// #[test_case("tests11.dat")]
```
Comment (CharlesChen0823) [inline tests/tree_construction.rs:66]: tests11.dat also pass, and there is no tests13.dat

**Review by CharlesChen0823 — APPROVED** [review]

**Review by Kiyoshika — APPROVED** [review]

File: src/html5/parser.rs [hunk 5]
```diff
@@ -268,1028 +274,1248 @@ impl<'stream> Html5Parser<'stream> {
...
+                    || name == "span"
+                    || name == "strong"
+                    || name == "strike"
+                    || name == "sub"
+                    || name == "sup"
+                    || name == "table"
+                    || name == "tt"
+                    || name == "u"
+                    || name == "ul"
+                    || name == "var" =>
```
Comment (emwalker) [inline src/html5/parser.rs:365]: )We should probably be using a match statement instead of an if statement here.)

**Review by emwalker — APPROVED** [review]

**Pull request merged**